test/performance/odp_ipsec_ordered.c

```diff
@@ -0,0 +1,1279 @@
+/** Global pointer to args */
+static args_t *gbl_args;
+
+#define MAX_WORKERS 128
+#define	NUM_CONFIGS		(sizeof(algs_config) / sizeof(ipsec_alg_config_t))
+
+/**
+ * Get the Result data for the calling thread
+ */
+static ipsec_run_result_t *cpu_get_result(void)
+{
+	int cpu = odp_cpu_id();
+
+	if (cpu >= MAX_WORKERS)
+		exit(EXIT_FAILURE);
+
+	return &gbl_args->result[cpu];
+}
+
+/**
+ * Initialize applicaiton parameters
+ */
+static void
+gbl_args_init(args_t *args)
+{
+	memset(args, 0, sizeof(args_t));
+
+	odp_atomic_init_u32(&args->exit_threads, 0);
+	odp_atomic_init_u32(&args->in_queue_count, 0);
+	odp_atomic_init_u32(&args->complete_cnt, 0);
+	odp_atomic_init_u32(&args->pkt_alloc, 0);
+	odp_atomic_init_u32(&args->pkt_free, 0);
+}
+
+/**
+ * Find corresponding config for given name. Returns NULL
+ * if config for given name is not found.
+ */
+static ipsec_alg_config_t *
+find_config_by_name(const char *name)
+{
+	unsigned int i;
+	ipsec_alg_config_t *ret = NULL;
+
+	for (i = 0; i < NUM_CONFIGS; i++) {
+		if (strcmp(algs_config[i].name, name) == 0) {
+			ret = algs_config + i;
+			break;
+		}
+	}
+	return ret;
+}
+
+/**
+ * Helper function that prints list of algorithms that this
+ * test understands.
+ */
+static void
+print_config_names(const char *prefix)
+{
+	unsigned int i;
+
+	for (i = 0; i < (sizeof(algs_config) / sizeof(ipsec_alg_config_t));
+			i++) {
+		printf("%s %s\n", prefix, algs_config[i].name);
+	}
+}
+
+/**
+ * Snap current time values and put them into 'rec'.
+ */
+static void
+fill_time_record(time_record_t *rec)
+{
+	gettimeofday(&rec->tv, NULL);
+	getrusage(RUSAGE_SELF, &rec->ru_self);
+	getrusage(RUSAGE_THREAD, &rec->ru_thread);
+}
+
+/**
+ * Calculated CPU time difference for given two rusage structures.
+ * Note it adds user space and system time together.
+ */
+static unsigned long long
+get_rusage_diff(struct rusage *start, struct rusage *end)
+{
+	unsigned long long rusage_diff;
+	unsigned long long rusage_start;
+	unsigned long long rusage_end;
+
+	rusage_start = (start->ru_utime.tv_sec * 1000000) +
+		(start->ru_utime.tv_usec);
+	rusage_start += (start->ru_stime.tv_sec * 1000000) +
+		(start->ru_stime.tv_usec);
+
+	rusage_end = (end->ru_utime.tv_sec * 1000000) +
+		(end->ru_utime.tv_usec);
+	rusage_end += (end->ru_stime.tv_sec * 1000000) +
+		(end->ru_stime.tv_usec);
+
+	rusage_diff = rusage_end - rusage_start;
+
+	return rusage_diff;
+}
+
+/**
+ * Get diff for RUSAGE_THREAD (current thread only) between two
+ * time snap records.
+ */
+static unsigned long long
+get_rusage_thread_diff(time_record_t *start, time_record_t *end)
+{
+	return get_rusage_diff(&start->ru_thread, &end->ru_thread);
+}
+
+/**
+ * Get diff of elapsed time between two time snap records
+ */
+static unsigned long long
+get_elapsed_usec(time_record_t *start, time_record_t *end)
+{
+	unsigned long long s;
+	unsigned long long e;
+
+	s = (start->tv.tv_sec * 1000000) + (start->tv.tv_usec);
+	e = (end->tv.tv_sec * 1000000) + (end->tv.tv_usec);
```

Comment:
These can overflow in 32 bit systems. You could fix that (for reasonable test durations) and also avoid the use of long long by first subtracting the seconds from each other and the microseconds from each other and then combining.

Reply:
It will be fixed in next version